<font size="+3" color='#053c96'><h2><center> Modeling</h2></center></font>
<figure>
<center><img src ="https://www.krasamo.com/wp-content/uploads/0730Building-ML-Models.jpg" width = "750" height = '600' alt="Classical Punk"/>
<font size="0" color='#053c96'><h4><center> Photo Credit: krasamo</h4></center></font>

<font size="+2" color='#053c96'><b> Contributor</b></font>  
<font size="+0" ><b> Umar Kabir</b></font>  

<a id='table-of-contents'></a>
[Table of Contents](#table-of-contents)

- [Modeling](#modeling)
  * [Data Transformation](#data-transformation)
  * [Model Selection](#model-selection)
  * [Model Training](#model-training)
  * [Model Evaluation](#model-evaluation)
  * [Hyperparameter Tuning](#hyperparameter-tuning)
- [Results](#results)
  * [Analysis Results](#analysis-results)
  * [Model Performance](#model-performance)
  * [Feature Importance](#feature-importance)
  * [Implications](#implications)
- [Conclusion](#conclusion)
  * [Summary](#summary)
  * [Limitations](#limitations)
  * [Recommendations](#recommendations)
- [References](#references)

<a id='importing-dependencies'></a>
<font size="+2" color='#053c96'><b> 1. Importing Libraries</b></font>  
[back to top](#table-of-contents)

In [1]:
import sys
# Insert the parent path relative to this notebook so we can import from the src folder.
sys.path.insert(0, "..")

from src.dependencies import *
from src.functions import *

<a id='data'></a>
<font size="+2" color='#053c96'><b> 2. Data</b></font>  
[back to top](#table-of-contents)  

The dataset was used for the well-known paper in genre classification "Musical genre classification of audio signals" by G. Tzanetakis and P. Cook in IEEE Transactions on Audio and Speech Processing 2002.

In [61]:
df = pd.read_csv("../data/processed_data.csv")

In [62]:
# Shuffle the data
df = shuffle(df)

In [63]:
df.head()

Unnamed: 0    Genre           Filename       Tempo    Complexity  \
904         904     rock     rock.00004.wav  103.359375  2.572210e-07   
275         275  country  country.00075.wav   95.703125 -1.603827e-07   
305         305    disco    disco.00005.wav  107.666016 -7.248804e-07   
958         958     rock     rock.00058.wav  135.999178  1.753696e-07   
85           85    blues    blues.00085.wav  143.554688 -9.480932e-08   

        Timbre  Spectral Energy  Temporal Energy  Beats Mean  \
904  21.528547     1.988146e+10       30041.7580  635.959184   
275  23.394657     1.541390e+10       23291.0840  605.822222   
305  18.063551     2.038556e+10       30520.9590  622.058824   
958  19.311353     3.440187e+09        5198.2754  637.426471   
85   24.034703     1.144591e+10       17295.2810  636.028169   

     Central Moments Mean  Zero Crossing Rate Mean  RMSE Mean  \
904              0.012449                 0.035807   0.151337   
275              0.010109                 0.090658   0.121423   
305              0.013015                 0.110514   0.180705   
958              0.001997                 0.093506   0.087335   
85               0.007198                 0.050781   0.085556   

     Spectral Contrast Mean  Spectral Roll-off Mean     MFCC Mean  \
904            66512.227156             4389.184570  49787.181998   
275            52381.338160             3987.231445   7156.856000   
305            33861.728885             4753.454590    157.576000   
958            47318.848190             6951.635742  20101.000101   
85              1256.782590             3002.087402      0.441937   

     Chroma Mean  Spectral Centroid Mean  Spectral Bandwidth Mean  \
904        0.588             1832.258174              2228.523490   
275       87.000             1957.229869              2068.278999   
305        6.350             2532.934124              2431.366786   
958        3.421             2913.545421              3125.952141   
85         7.293             1315.205904              1812.502520   

     Spectral Flatness Mean  Loudness Mean  Harmonicity Mean  \
904                0.005341     -16.702490          0.073103   
275                0.003763     -18.371674          0.040451   
305                0.020712     -15.059725          0.037915   
958                0.020351     -21.224725         -0.062207   
85                 0.001176     -21.447978         -0.032428   

     Percussiveness Mean   Beats std  Central Moments std  \
904             0.033637  354.140060             0.019297   
275             0.010656  348.728120             0.014648   
305             0.026055  351.209027             0.019059   
958            -0.017444  367.730333             0.003377   
85             -0.008333  370.120269             0.011012   

     Zero Crossing Rate std  RMSE std  Spectral Contrast std  \
904                0.023376  0.037688                    0.0   
275                0.035674  0.013714                    0.0   
305                0.033480  0.038674                    0.0   
958                0.028375  0.009161                    0.0   
85                 0.026125  0.012299                    0.0   

     Spectral Roll-off std  MFCC std  Chroma std  Spectral Centroid std  \
904            1829.336821       0.0         0.0             678.522603   
275             308.099597       0.0         0.0             348.732212   
305             393.766772       0.0         0.0             158.319354   
958             471.550381       0.0         0.0             238.402583   
85             2102.562430       0.0         0.0             640.388542   

     Spectral Bandwidth std  Spectral Flatness std  Loudness std  \
904              562.165603               0.004583      2.356688   
275              101.466110               0.001420      1.011701   
305               32.471064               0.003537      1.863141   
958              189.320929               0.007267      0.921779   
85               584.959547     

In [64]:
X = df.drop(['Genre', 'Filename', 'Unnamed: 0', 'Spectral Contrast zscore', 'MFCC zscore',
            'Chroma zscore', 'Spectral Contrast kurtosis', 'MFCC kurtosis',
            'Chroma kurtosis', 'Spectral Contrast skew', 'MFCC skew',
            'Chroma skew',], axis=1)
y = df['Genre']

<a id='modeling'></a>

<font size="+2" color='#053c96'><b> 4. Modeling</b></font>  
[back to top](#table-of-contents)

<a id='data-transformation'></a>
<font size="+0" color='green'><b> Feature Selection</b></font>  
[back to top](#table-of-contents)
<a id='modeling'></a>

In [65]:
# Initialize the OneHotEncoder
encoder = LabelEncoder()

# Fit and transform the 'Genre' column
y_encoded = encoder.fit_transform(y)

In [131]:
# Apply Min-Max scaling to make the features non-negative
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Initialize the feature selector
selector = SelectKBest(score_func=chi2, k=50)  # Select top 10 features based on chi-squared test

# Apply feature selection
X_selected = selector.fit_transform(X_scaled, y)

# Get the indices of the selected features
selected_indices = selector.get_support(indices=True)

# Get the names of the selected features
selected_features = X.columns[selected_indices]

In [132]:
# Create a logistic regression classifier
classifier = LogisticRegression()

# Initialize the RFE selector
rfe_selector = RFE(estimator=classifier, n_features_to_select=50)  # Select top 5 features

# Apply feature selection
X_selected = rfe_selector.fit_transform(X, y)

# Get the indices of the selected features
selected_indices = rfe_selector.get_support(indices=True)

# Get the names of the selected features
selected_features1 = X.columns[selected_indices]


In [133]:
# Create a logistic regression classifier with L1 regularization (Lasso)
classifier = LogisticRegression(penalty='l1', solver='liblinear')

# Initialize the feature selector
selector = SelectFromModel(estimator=classifier, max_features=50)

# Apply feature selection
X_selected = selector.fit_transform(X, y)

# Get the indices of the selected features
selected_indices = selector.get_support(indices=True)

# Get the names of the selected features
selected_features2 = X.columns[selected_indices]

In [134]:
# Perform feature selection using mutual information
selector = SelectKBest(score_func=mutual_info_classif, k=50)  # Select top 5 features based on mutual information

# Apply feature selection
X_selected = selector.fit_transform(X, y)

# Get the indices of the selected features
selected_indices = selector.get_support(indices=True)

# Get the names of the selected features
selected_features3 = X.columns[selected_indices]


In [135]:
features_ = list(set(selected_features) | set(selected_features1) | set(selected_features2) | set(selected_features3))
print(features_)

['Loudness max', 'Spectral Bandwidth sum', 'Temporal Energy', 'Beats sum', 'Spectral Bandwidth iqr', 'Spectral Bandwidth skew', 'Beats skew', 'Spectral Roll-off skew', 'Central Moments var', 'Zero Crossing Rate sum', 'Beats std', 'Spectral Roll-off std', 'Spectral Contrast sum', 'Harmonicity sum', 'Harmonicity std', 'Central Moments kurtosis', 'MFCC sum', 'Spectral Flatness iqr', 'Spectral Roll-off var', 'MFCC Mean', 'Spectral Bandwidth std', 'Zero Crossing Rate kurtosis', 'Central Moments skew', 'Chroma max', 'RMSE var', 'Zero Crossing Rate min', 'Chroma min', 'Tempo', 'Chroma Mean', 'Central Moments iqr', 'Spectral Centroid sum', 'Loudness kurtosis', 'Harmonicity kurtosis', 'Spectral Flatness var', 'Loudness std', 'Spectral Contrast min', 'Spectral Roll-off iqr', 'RMSE iqr', 'Spectral Centroid Mean', 'Spectral Centroid max', 'Harmonicity skew', 'Spectral Flatness max', 'Beats Mean', 'Spectral Roll-off min', 'Spectral Contrast Mean', 'Spectral Flatness min', 'Percussiveness skew', 'RM

<a id='data-transformation'></a>
<font size="+0" color='green'><b> Data Transformation</b></font>  
[back to top](#table-of-contents)
<a id='modeling'></a>

In [136]:
# Calculate the mean and standard deviation for each column
mean = X.mean()
std = X.std()

# Perform Z-score normalization on the entire dataframe
train_df = (X - mean) / std
train_df = train_df.fillna(0)

In [137]:
# Applying natural logarithm transformation to all columns
train_df_transformed = X.apply(np.log)
# Replace all NaN values with 0
train_df_transformed = train_df_transformed.fillna(0)
# Replace all inf values with 0
train_df_transformed = train_df_transformed.replace([np.inf, -np.inf], 0)

In [138]:
# Perform standardization on the train dataframe
scaler = StandardScaler()
train_df_scaled = scaler.fit_transform(X)

# Create a PCA object
pca = PCA()

# Apply PCA on the scaled train dataframe
train_df_pca = pca.fit_transform(train_df_scaled)

# Create a new dataframe with the principal components
train_df_pca = pd.DataFrame(data=train_df_pca, columns=['PC{}'.format(i+1) for i in range(pca.n_components_)])

In [139]:
#Pareto Principle Split
X_train, X_test, y_train, y_test = train_test_split(train_df_transformed[features_], y_encoded, test_size=0.3, random_state=42)

<a id='model-selection'></a>

<font size="+1" color='#780404'><b> 4.1 Model Selection</b></font>  
[back to top](#table-of-contents)

In [154]:
# Multiclass Support Vector Machines (SVM)
svm_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', OneVsRestClassifier(SVC(kernel='rbf', C=1, gamma='scale')))
])

# K-Means Clustering
kmeans_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('kmeans', KMeans(n_clusters=10))
])

# K-Nearest Neighbors (KNN)
knn_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('knn', KNeighborsClassifier(n_neighbors=5))
])

# Feedforward Neural Network
cnn_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', Sequential([
        Dense(64, activation='relu', input_shape=(len(features_),)),
        Dense(32, activation='relu'),
        Dense(10, activation='softmax')
    ]))
])
epochs = 50
batch_size = 32

# Multiclass Logistic Regression
logreg_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', LogisticRegression(max_iter=1000))
])

# Multinomial Naive Bayes
nb_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', MultinomialNB())
])

# Random Forest
rf_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', RandomForestClassifier(n_estimators=100, random_state=42))
])

# Gradient Boosting
gb_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', GradientBoostingClassifier(n_estimators=100, random_state=42))
])

# Linear Support Vector Machines (SVM)
linear_svm_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', LinearSVC())
])

# AdaBoost Classifier
adaboost_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', AdaBoostClassifier(n_estimators=100, random_state=42))
])

# XGBoost Classifier
xgb_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', XGBClassifier(n_estimators=100, random_state=42))
])

# Multi-layer Perceptron (MLP) Classifier
mlp_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', MLPClassifier(hidden_layer_sizes=(64, 32), max_iter=1000, random_state=42))
])

# Gaussian Process Classifier
gaussian_process_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', GaussianProcessClassifier(random_state=42))
])

# LightGBM Classifier
lgbm_pipeline = Pipeline([
    ('clf', LGBMClassifier(n_estimators=100, random_state=42))
])

# CatBoost Classifier
catboost_pipeline = Pipeline([
    ('clf', CatBoostClassifier(iterations=100, random_seed=42, verbose=0))
])

# Quadratic Discriminant Analysis (QDA)
qda_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', QuadraticDiscriminantAnalysis())
])

# Ridge Classifier
ridge_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', RidgeClassifier(alpha=1.0))
])

# Perceptron
perceptron_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', Perceptron(max_iter=1000, random_state=42))
])


<a id='model-training'></a>

<font size="+1" color='#780404'><b> 4.2 Model Training</b></font>  
[back to top](#table-of-contents)

In [157]:
# Fit SVM pipeline
svm_pipeline.fit(X_train, y_train)

# Fit KNN pipeline
knn_pipeline.fit(X_train, y_train)

# Fit Logistic Regression pipeline
logreg_pipeline.fit(X_train, y_train)

# Fit Multinomial Naive Bayes pipeline
#nb_pipeline.fit(X_train, y_train)

# Fit Random Forest pipeline
rf_pipeline.fit(X_train, y_train)

# Fit Gradient Boosting pipeline
gb_pipeline.fit(X_train, y_train)

# Fit Linear SVM pipeline
linear_svm_pipeline.fit(X_train, y_train)

# Fit AdaBoost pipeline
adaboost_pipeline.fit(X_train, y_train)

# Fit XGBoost pipeline
xgb_pipeline.fit(X_train, y_train)

# Fit Multi-layer Perceptron (MLP) pipeline
mlp_pipeline.fit(X_train, y_train)

# Fit Gaussian Process pipeline
gaussian_process_pipeline.fit(X_train, y_train)

# Fit LightGBM pipeline
lgbm_pipeline.fit(X_train, y_train)

# Fit CatBoost pipeline
catboost_pipeline.fit(X_train, y_train)

# Fit Quadratic Discriminant Analysis (QDA) pipeline
qda_pipeline.fit(X_train, y_train)

# Fit Ridge Classifier pipeline
ridge_pipeline.fit(X_train, y_train)

# Fit Perceptron pipeline
perceptron_pipeline.fit(X_train, y_train)

Epoch 1/50
18/18 [==============================] - 1s 19ms/step - loss: 1.1374 - accuracy: 0.6232 - val_loss: 2.2781 - val_accuracy: 0.2500
Epoch 2/50
18/18 [==============================] - 0s 3ms/step - loss: 1.0811 - accuracy: 0.6268 - val_loss: 2.4104 - val_accuracy: 0.2143
Epoch 3/50
18/18 [==============================] - 0s 3ms/step - loss: 1.0770 - accuracy: 0.6339 - val_loss: 2.3618 - val_accuracy: 0.2429
Epoch 4/50
18/18 [==============================] - 0s 3ms/step - loss: 1.0521 - accuracy: 0.6500 - val_loss: 2.3834 - val_accuracy: 0.2500
Epoch 5/50
18/18 [==============================] - 0s 3ms/step - loss: 1.0489 - accuracy: 0.6393 - val_loss: 2.3683 - val_accuracy: 0.2643
Epoch 6/50
18/18 [==============================] - 0s 3ms/step - loss: 1.0468 - accuracy: 0.6464 - val_loss: 2.4521 - val_accuracy: 0.2357
Epoch 7/50
18/18 [==============================] - 0s 3ms/step - loss: 1.0289 - accuracy: 0.6661 - val_loss: 2.3646 - val_accuracy: 0.2286
Epoch 8/50
18/18 [=

Pipeline(steps=[('scaler', StandardScaler()),
                ('clf', Perceptron(random_state=42))])

<a id='model-evaluation'></a>

<font size="+1" color='#780404'><b> 4.3 Model Evaluation</b></font>  
[back to top](#table-of-contents)

In [159]:
# Initialize the lists for performance metrics and confusion matrices
f1_list = []
pr_list = []
rc_list = []
acc_list = []

# Evaluate additional models
# evaluate the SVM pipeline

y_pred_svm = svm_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_svm, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_svm, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_svm, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_svm))


# evaluate the KNN pipeline
y_pred_knn = knn_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_knn, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_knn, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_knn, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_knn))


# Evaluate Logistic Regression pipeline
y_pred_logreg = logreg_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_logreg, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_logreg, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_logreg, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_logreg))


# Evaluate Random Forest pipeline
y_pred_rf = rf_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_rf, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_rf, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_rf, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_rf))

# Evaluate Gradient Boosting pipeline
y_pred_gb = gb_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_gb, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_gb, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_gb, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_gb))

# Evaluate Linear SVM pipeline
y_pred_linear_svm = linear_svm_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_linear_svm, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_linear_svm, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_linear_svm, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_linear_svm))

# Evaluate AdaBoost pipeline
y_pred_adaboost = adaboost_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_adaboost, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_adaboost, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_adaboost, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_adaboost))

# Evaluate XGBoost pipeline
y_pred_xgb = xgb_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_xgb, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_xgb, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_xgb, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_xgb))

# Evaluate Multi-layer Perceptron (MLP) pipeline
y_pred_mlp = mlp_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_mlp, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_mlp, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_mlp, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_mlp))

# Evaluate Gaussian Process pipeline
y_pred_gaussian_process = gaussian_process_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_gaussian_process, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_gaussian_process, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_gaussian_process, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_gaussian_process))

# Evaluate LightGBM pipeline
y_pred_lgbm = lgbm_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_lgbm, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_lgbm, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_lgbm, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_lgbm))

# Evaluate CatBoost pipeline
y_pred_catboost = catboost_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_catboost, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_catboost, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_catboost, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_catboost))

# Evaluate Quadratic Discriminant Analysis (QDA) pipeline
y_pred_qda = qda_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_qda, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_qda, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_qda, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_qda))

# Evaluate Ridge Classifier pipeline
y_pred_ridge = ridge_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_ridge, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_ridge, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_ridge, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_ridge))

# Evaluate Perceptron pipeline
y_pred_perceptron = perceptron_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_perceptron, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_perceptron, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_perceptron, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_perceptron))



In [160]:
# Create a list of model names
model_list = ['SVM', 'KNN', 'Logistic Regression', 'Random Forest', 'Gradient Boosting',
            'Linear SVM', 'AdaBoost', 'XGBoost', 'Multi-layer Perceptron', 'Gaussian Process',
            'LightGBM', 'CatBoost', 'QDA', 'Ridge Classifier', 'Perceptron']

In [161]:
result = pd.DataFrame({'Model': model_list, 'F1': f1_list, 'Precision': pr_list, 'Recall': rc_list, 'Accuracy': acc_list})

In [162]:
result

Model        F1  Precision    Recall  Accuracy
0                      SVM  0.404171   0.400079  0.430000  0.430000
1                      KNN  0.314228   0.322350  0.320000  0.320000
2      Logistic Regression  0.393931   0.393724  0.396667  0.396667
3            Random Forest  0.425867   0.425307  0.433333  0.433333
4        Gradient Boosting  0.488082   0.491550  0.490000  0.490000
5               Linear SVM  0.386867   0.379972  0.400000  0.400000
6                 AdaBoost  0.211073   0.246405  0.276667  0.276667
7                  XGBoost  0.487744   0.495255  0.496667  0.496667
8   Multi-layer Perceptron  0.345478   0.356324  0.346667  0.346667
9         Gaussian Process  0.260263   0.293756  0.246667  0.246667
10                LightGBM  0.503926   0.514287  0.506667  0.506667
11                CatBoost  0.443086   0.455901  0.436667  0.436667
12                     QDA  0.082254   0.363841  0.096667  0.096667
13        Ridge Classifier  0.375370   0.374564  0.390000  0.390000
14              Perceptron  0.276960   0.295416  0.273333  0.273333

<a id='hyperparameter-tuning'></a>
<font size="+1" color='#780404'><b> 4.4 Hyperparameter Tuning</b></font>  
[back to top](#table-of-contents)

<a id='results'></a>
<font size="+2" color='#053c96'><b> 5. Results</b></font>  
[back to top](#table-of-contents)

<a id='analysis-results'></a>

<font size="+1" color='#780404'><b> 5.1 Analysis Results</b></font>  
[back to top](#table-of-contents)

<a id='model-performance'></a>

<font size="+1" color='#780404'><b> 5.2 Model Performance</b></font>  
[back to top](#table-of-contents)

<a id='feature-importance'></a>

<font size="+1" color='#780404'><b> 5.3 Feature Importance</b></font>  
[back to top](#table-of-contents)

<a id='implications'></a>

<font size="+1" color='#780404'><b> 5.4 Implications</b></font>  
[back to top](#table-of-contents)

<a id='conclusion'></a>

<font size="+2" color='#053c96'><b> 6. Conclusion</b></font>  
[back to top](#table-of-contents)

<a id='summary'></a>

<font size="+1" color='#780404'><b> 6.1 Summary</b></font>  
[back to top](#table-of-contents)

<a id='limitations'></a>

<font size="+1" color='#780404'><b> 6.2 Limitations</b></font>  
[back to top](#table-of-contents)

<a id='recommendations'></a>

<font size="+1" color='#780404'><b> 6.3 Recommendations</b></font>  
[back to top](#table-of-contents)

<a id='references'></a>

<font size="+2" color='#053c96'><b> 7. References</b></font>  
[back to top](#table-of-contents)